implemneted feature 'add me to your friends' https://github.com/solid/profile-pane/issues/8

theRealImy opened the pull request:
Added the new 'Add me to your friends' button under profile (see screenshot).
![image](https://user-images.githubusercontent.com/4144203/130831076-da5faf8c-1118-4405-a421-ea804d454479.png)
The test case with logged in user will change when auth changes.

theRealImy commented:
I've improved the code based on the feedback.
However to keep UI consistent with the 'Chat with me' button which is also displayed on the profile pane, i did not add a greyed out button but followed same design like 'Chat with me'. I propose revising button design afterwards, for all buttons. 
For now:
* if one is not logged in the button is like the 'chat with me' without filling
<img width="269" alt="Screenshot 2021-09-01 at 14 47 04" src="https://user-images.githubusercontent.com/4144203/131819969-00fb493d-b000-40f4-a3f6-79792870eb5e.png">

* the button is active and if pressed one gets:
<img width="290" alt="Screenshot 2021-09-02 at 11 33 01" src="https://user-images.githubusercontent.com/4144203/131820288-864a4a51-55e5-4354-ae80-ea4b99eeccf5.png">

* if one is logged in and friend does not exist, the button get a filling like:
<img width="278" alt="Screenshot 2021-09-01 at 14 46 43" src="https://user-images.githubusercontent.com/4144203/131820049-a647cd56-75c5-4db2-9e9c-6b8b3b81ec3a.png">

* if friend is added a message appears:
<img width="284" alt="Screenshot 2021-09-02 at 11 34 29" src="https://user-images.githubusercontent.com/4144203/131820498-d7832650-b3fa-4c3a-904a-a6989de1afb0.png">

* if I log in and I already have this friend in my list I see a different message on the button:
<img width="291" alt="Screenshot 2021-09-01 at 14 45 30" src="https://user-images.githubusercontent.com/4144203/131820619-e825ccbe-7ec8-4eab-9b84-5570f68cf3a6.png">

* If I still press the button even if 'Already part of friend list' I get a message like:
<img width="297" alt="Screenshot 2021-09-01 at 14 46 01" src="https://user-images.githubusercontent.com/4144203/131820697-7f7e429f-3289-48ea-9d14-e5b1ea5cae9c.png">

Part missing which I am not sure if it should work: button design is not refreshed after clicking it, should the UI have this behaviour? Also, I noticed that the data is not really updated real-time.
Example: 
*I add a new friend
* I go to my pod and delete the friend again (manually)
* I try to add the friend again (but I did not refresh the page yet) 
* friend is not added.  -> Code wise, I do a store.fetcher.load(webId) but it still shows that the dedicated ns.foaf('knows') statement exists, so friend is not added because it shows it exists already. Not sure if store.fetcher.load should work... or am I missing another function call in between?

SharonStrats commented:
Ideas: I'm wondering in the case the user is not logged in, could the button say something like "login to add me to your friends"

SharonStrats commented on src/addMeToYourFriends.ts:
```diff
@@ -0,0 +1,146 @@
+//TODO create dedicated component in UI
+const positiveFrontendMessageDiv = <HTMLDivElement>document.createElement("div");
+positiveFrontendMessageDiv.setAttribute(
+  "style",
```
In this case maybe it's appropriate to add a message style to baseStyles and add it in the same way as styles = .. styleMap.  If there are specific styles to friends you can also create another file called perhaps friendStyles.

SharonStrats replied:
Here, see @timbl comment in Gitter regarding the styles should be from solid-ui and use setStyle().  I will see if I can find an example and add here.  The file to find them in is solid-ui/src/style.js and you can add a style here that you need. I see something called messageBodyStyle...  this is something to look at although I don't know if it's exactly what you need.  I will do some more research for this as well, as we definitely have errors in other places. I'm going to make another ticket though for styles in general as to make it easier to read in solid-ui I think we should perhaps have more than one style.js file going forward.

SharonStrats replied:
I actually couldn't find many examples but here is one    ` head = table.appendChild(dom.createElement('tr'));
            style.setStyle(head, 'autocompleteRowStyle'); // textInputStyle or`.  I also think solid-ui/src/messageArea.js is worth checking out for error messages.  I can't be sure about this but something to look into.

theRealImy replied:
regarding styles, i did not change so much because it will cause inconsistencies on the Profile Pane. I propose opening an additional ticket with topic - align profile pane styles with UI.

SharonStrats commented on src/addMeToYourFriends.ts:
```diff
@@ -0,0 +1,146 @@
+
+//Same not logged in error message like on 'Chat with me' buton
+const userNotLoggedInErrorMessage = "Current user not found! Not logged in?";
+const internalErrorMessage = "An internal error occured!";
```
I think I would pop the messages into another file as well.  this was originally a small pane, but as we are adding more functionality we should think about having folders.  So for instance you could create a friend folder under src and pop all the friend functionality in to there and then this message file would be part of that.

theRealImy replied:
I created a texts file for all messages but I did not add it to its own folder because I gathered also 2 messages from the Chat code. (so a text file for all texts in profile pane )

SharonStrats commented on src/addMeToYourFriends.ts:
```diff
@@ -0,0 +1,146 @@
+const addMeToYourFriendsButtonText = "Add me to your friend list";
+const friendExistsAlreadyButtonText = "Already part of friend list";
+
+const addMeToYourFriendsHtml = (
```
Suggestion: I don't think you need to have the html.  you might call it the thing it is like addMeToYourFriendsButton perhaps.

theRealImy replied:
I call it Div instead of Html

SharonStrats commented on src/addMeToYourFriends.ts:
```diff
@@ -0,0 +1,146 @@
+import * as UI from "solid-ui";
+import { padding, textCenter } from "./baseStyles";
+
+let buttonContainer = <HTMLDivElement>document.createElement("div");
```
praise: great typing throughout.

SharonStrats commented on src/addMeToYourFriends.ts:
```diff
@@ -0,0 +1,146 @@
+    }
+  );
+  //this is to make clear which style we have, for code readability
+  button.setAttribute("class", "textButton-0-1-3"); //style of 'Primary' UI button with needsBorder=true
```
praise: great idea!

SharonStrats commented on src/addMeToYourFriends.ts:
```diff
@@ -0,0 +1,146 @@
+    "click",
+    () =>
+      saveNewFriend(subject, context)
+        .then(() => reportToFrontend(true, context, friendWasAddedSuccesMessage))
```
suggestion: I think I've seen notify before versus reportTo. perhaps have a quick look and if so could potentially change just for consistency.  I'm not 100% about this.  just a thought.

theRealImy replied:
I only found announce and mention and I renamed my functions to these.

SharonStrats commented on src/addMeToYourFriends.ts:
```diff
@@ -0,0 +1,146 @@
+  else return true;
+}
+
+function reportToFrontend(positive: boolean, context: DataBrowserContext, message: string) {
```
suggestion: if you do create the friend folder, then this could be in a helper file within that folder.

theRealImy replied:
done

SharonStrats reviewed (changes requested):
Really great work!  So exciting!

some additional more general suggestions.  Right now there don't seem to be unit tests in general, but for this I think it would be good to have them.  Just for your helper functions.  So if you do create the folder friends and you have the helper.ts, you could just have one unit test file helper.test.ts to test your helper functions.

Awesome work 🥳!

theRealImy commented:
> Really great work! It's working very nicely. I will make further comments on your branch in your repository if I find code suggestions.
> 
> Ideas: I'm wondering in the case the user is not logged in, could the button say something like "login to add me to your friends".
> 
> it should be updated realtime. I will do some research on this, because I don't know off the top of my head how this happens. But it does happen with addresses for instance. If you click to add a new contact, it will pop back and tell you it was created. Actually I see what you mean, it changes somewhere else and do the changes take effect without a refresh... hmm I'm not sure about this.

theRealImy commented:
> Really great work! It's working very nicely. I will make further comments on your branch in your repository if I find code suggestions.
> 
> Ideas: I'm wondering in the case the user is not logged in, could the button say something like "login to add me to your friends".
> 
> it should be updated realtime. I will do some research on this, because I don't know off the top of my head how this happens. But it does happen with addresses for instance. If you click to add a new contact, it will pop back and tell you it was created. Actually I see what you mean, it changes somewhere else and do the changes take effect without a refresh... hmm I'm not sure about this.

I updated the button text as mentioned. 
I also added more tests.
Regarding refresh - this is a TODO

theRealImy commented:
I changed the button message as follows:
<img width="245" alt="Screenshot 2021-09-14 at 16 31 01" src="https://user-images.githubusercontent.com/4144203/133277294-4071b986-b147-479c-a062-9e8c3b7e0090.png">
<img width="239" alt="Screenshot 2021-09-14 at 16 31 16" src="https://user-images.githubusercontent.com/4144203/133277306-60df3e97-ebda-4508-bcac-04927d79b172.png">
<img width="246" alt="Screenshot 2021-09-14 at 16 31 49" src="https://user-images.githubusercontent.com/4144203/133277322-52f30eaa-bb11-4ca8-924b-9ae0c5bace59.png">

theRealImy commented:
refresh also implemented now
